example/src/main/java/io/netty/example/http2/tiles/Http2RequestHandler.java

```diff
@@ -0,0 +1,120 @@
+
+package io.netty.example.http2.tiles;
+
+import static io.netty.buffer.Unpooled.EMPTY_BUFFER;
+import static io.netty.example.http2.Http2ExampleUtil.firstValue;
+import static io.netty.example.http2.Http2ExampleUtil.toInt;
+import static io.netty.handler.codec.http.HttpHeaderNames.CONTENT_TYPE;
+import static io.netty.handler.codec.http.HttpHeaderUtil.setContentLength;
+import static io.netty.handler.codec.http.HttpResponseStatus.BAD_REQUEST;
+import static io.netty.handler.codec.http.HttpResponseStatus.OK;
+import static io.netty.handler.codec.http.HttpVersion.HTTP_1_1;
+import static java.lang.Integer.valueOf;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.SimpleChannelInboundHandler;
+import io.netty.handler.codec.http.DefaultFullHttpResponse;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.FullHttpResponse;
+import io.netty.handler.codec.http.QueryStringDecoder;
+import io.netty.handler.codec.http2.HttpUtil;
+import io.netty.handler.codec.http2.InboundHttp2ToHttpAdapter;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Handles all the requests for data. It receives a {@link FullHttpRequest},
+ * which has been converted by a {@link InboundHttp2ToHttpAdapter} before it
+ * arrived here. For further details, check {@link Http2OrHttpHandler} where the
+ * pipeline is setup.
+ */
+public class Http2RequestHandler extends SimpleChannelInboundHandler<FullHttpRequest> {
+
+    private static final String LATENCY = "latency";
+    private static final int MAX_LATENCY = 1000;
+    private static final String IMAGE_COORDINATE_Y = "y";
+    private static final String IMAGE_COORDINATE_X = "x";
+
+    @Override
+    protected void channelRead0(ChannelHandlerContext ctx, FullHttpRequest request) throws Exception {
+        QueryStringDecoder queryString = new QueryStringDecoder(request.uri());
+        String streamId = streamId(request);
+        int latency = toInt(firstValue(queryString, LATENCY), 0);
+        if (latency > MAX_LATENCY) {
+            sendBadRequest(ctx, streamId);
+            return;
+        }
+        String x = firstValue(queryString, IMAGE_COORDINATE_X);
+        String y = firstValue(queryString, IMAGE_COORDINATE_Y);
+        if (x == null || y == null) {
+            handlePage(ctx, streamId, latency);
+        } else {
+            handleImage(x, y, ctx, streamId, latency);
+        }
+    }
+
+    private void sendBadRequest(ChannelHandlerContext ctx, String streamId) {
+        FullHttpResponse response = new DefaultFullHttpResponse(HTTP_1_1, BAD_REQUEST, EMPTY_BUFFER);
+        streamId(response, streamId);
+        ctx.writeAndFlush(response);
+    }
+
+    private void handleImage(String x, String y, ChannelHandlerContext ctx, String streamId, int latency) {
+        ByteBuf image = ImageCache.INSTANCE.image(valueOf(x), valueOf(y));
+        FullHttpResponse response = new DefaultFullHttpResponse(HTTP_1_1, OK, image);
```

Comment:
It seems so, however it is working and I haven't had issues with the indexes, so I tried to dig a bit... apparently, [it won't touch the readerIndex if it manages to write the full content of the buffer in one shot](https://github.com/netty/netty/blob/85ec4d9cc4de1ce5db810ada44f43528785a6d43/transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java#L317). Index will [get updated if it's a partial write tough](https://github.com/netty/netty/blob/85ec4d9cc4de1ce5db810ada44f43528785a6d43/transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java#L337).